File: import_engine.py
"""Generator-based import engine for the web UI.

Wraps the existing lib/ modules into a generator that yields structured
progress events. Each event is a dict that the Flask app converts to
an SSE data line for the browser.

This module does NOT modify any existing lib/ code — it imports and
calls the same functions the CLI uses.
"""

import logging
from pathlib import Path
from typing import Generator

import yaml
from jinja2 import Environment, FileSystemLoader

from lib.confluence_client import ConfluenceClient
from lib.content_converter import convert_content
from lib.date_utils import (
    format_display_date,
    format_flare_filename,
    format_image_folder,
    format_overview_link_text,
    format_toc_title,
    parse_confluence_title,
)
from lib.flare_updater import update_mini_toc, update_overview_page
from lib.image_handler import create_image_folder, download_and_place_images

logger = logging.getLogger(__name__)

# BOM character for Flare HTM files
BOM = "\ufeff"


def validate_config(config_path: Path) -> dict:
    """Load and validate config.yaml, returning result dict.

    Unlike the CLI's load_config(), this does NOT call sys.exit().
    Returns {"valid": True, "config": {...}} or {"valid": False, "error": "..."}.
    """
    if not config_path.exists():
        return {
            "valid": False,
            "error": "Config file not found. Copy config.yaml.example to config.yaml and fill in your credentials.",
        }

    try:
        with open(config_path, encoding="utf-8") as f:
            config = yaml.safe_load(f)
    except Exception as e:
        return {"valid": False, "error": f"Failed to parse config.yaml: {e}"}

    conf = config.get("confluence", {})
    for field in ("base_url", "email", "api_token", "production_parent_id"):
        if not conf.get(field):
            return {
                "valid": False,
                "error": f"Missing required field: confluence.{field}",
            }

    flare = config.get("flare_project", {})
    if not flare.get("root"):
        return {
            "valid": False,
            "error": "Missing required field: flare_project.root",
        }

    return {"valid": True, "config": config}


def resolve_paths(config: dict, script_dir: Path) -> dict:
    """Resolve all Flare project paths relative to script directory.

    Same logic as confluence_to_flare.resolve_flare_paths().
    """
    flare = config["flare_project"]
    root = (script_dir / flare["root"]).resolve()

    return {
        "root": root,
        "release_notes_dir": root / flare.get(
            "release_notes_dir",
            "Content/E_Landing Topics/Release Notes",
        ),
        "images_dir": root / flare.get(
            "images_dir",
            "Content/Resources/From Confluence/Release Notes",
        ),
        "overview_file": root / flare.get(
            "overview_file",
            "Content/E_Landing Topics/Release Notes/Release Notes Overview.htm",
        ),
        "toc_file": root / flare.get(
            "toc_file",
            "Project/TOCs/Landing Topic Mini TOCs/Visualization Guide Mini-TOC.fltoc",
        ),
    }


def get_existing_releases(release_notes_dir: Path) -> set[str]:
    """Get filenames of existing release note HTM files."""
    if not release_notes_dir.exists():
        return set()
    return {f.name for f in release_notes_dir.glob("Release Notes *.htm")}


def create_client(config: dict) -> ConfluenceClient:
    """Create an authenticated Confluence client from config."""
    conf = config["confluence"]
    return ConfluenceClient(conf["base_url"], conf["email"], conf["api_token"])


def run_import(
    page_ids: list[str],
    config: dict,
    script_dir: Path,
    force: bool = False,
) -> Generator[dict, None, None]:
    """Run the import pipeline, yielding progress events.

    Args:
        page_ids: List of Confluence page IDs to import.
        config: Validated config dict (from validate_config).
        script_dir: Path to the confluence_to_flare directory.
        force: If True, overwrite existing Flare files.

    Yields:
        Progress event dicts with "type" and other fields.
    """
    total = len(page_ids)
    yield {"type": "start", "total": total, "message": f"Starting import of {total} page(s)"}

    # Setup
    try:
        paths = resolve_paths(config, script_dir)
        if not paths["root"].exists():
            yield {"type": "error", "index": -1, "message": f"Flare project not found at {paths['root']}"}
            yield {"type": "complete", "imported": 0, "skipped": 0, "errors": 1, "message": "Import failed"}
            return

        client = create_client(config)
        template_dir = script_dir / "templates"
        env = Environment(loader=FileSystemLoader(str(template_dir)))
        template = env.get_template("release_note.htm.j2")
        existing = get_existing_releases(paths["release_notes_dir"])
    except Exception as e:
        yield {"type": "error", "index": -1, "message": f"Setup failed: {e}"}
        yield {"type": "complete", "imported": 0, "skipped": 0, "errors": 1, "message": "Import failed"}
        return

    imported_count = 0
    skipped_count = 0
    error_count = 0

    for i, page_id in enumerate(page_ids):
        try:
            # Fetch title
            title = client.get_page_title(page_id)
            yield {
                "type": "page_start",
                "index": i,
                "page_id": page_id,
                "title": title,
                "message": f"Importing: {title}",
            }

            # Parse date
            dt = parse_confluence_title(title)
            if not dt:
                yield {"type": "error", "index": i, "message": f"Could not parse date from title: {title}"}
                error_count += 1
                continue

            filename = format_flare_filename(dt)
            display_date = format_display_date(dt)
            image_folder_name = format_image_folder(dt)

            # Check if already exists
            if filename in existing and not force:
                yield {"type": "skip", "index": i, "filename": filename, "message": f"Already exists: {filename}"}
                skipped_count += 1
                continue

            # Step 1: Download images
            yield {"type": "step", "index": i, "step": "images", "message": "Downloading images..."}
            image_dest = create_image_folder(paths["images_dir"], image_folder_name)
            image_mapping = download_and_place_images(client, page_id, image_dest)
            img_count = len(image_mapping)
            yield {"type": "step", "index": i, "step": "images_done", "message": f"Downloaded {img_count} image(s)"}

            # Step 2: Fetch and convert content
            yield {"type": "step", "index": i, "step": "content", "message": "Converting page content..."}
            xhtml = client.get_page_content(page_id)
            body_content = convert_content(xhtml, image_mapping, image_folder_name)
            yield {"type": "step", "index": i, "step": "content_done", "message": "Converted page content"}

            # Step 3: Render template
            yield {"type": "step", "index": i, "step": "render", "message": "Rendering Flare topic..."}
            htm = template.render(
                display_date=display_date,
                body_content=body_content,
            )
            yield {"type": "step", "index": i, "step": "render_done", "message": "Rendered Flare topic"}

            # Step 4: Write HTM file
            yield {"type": "step", "index": i, "step": "write", "message": f"Writing {filename}..."}
            dest_file = paths["release_notes_dir"] / filename
            dest_file.parent.mkdir(parents=True, exist_ok=True)
            dest_file.write_text(BOM + htm, encoding="utf-8")
            yield {"type": "step", "index": i, "step": "write_done", "message": f"Wrote {filename}"}

            # Step 5: Update Overview page
            yield {"type": "step", "index": i, "step": "overview", "message": "Updating Overview page..."}
            update_overview_page(
                paths["overview_file"],
                filename,
                format_overview_link_text(dt),
                release_date=dt,
            )
            yield {"type": "step", "index": i, "step": "overview_done", "message": "Updated Overview page"}

            # Step 6: Update Mini-TOC
            yield {"type": "step", "index": i, "step": "toc", "message": "Updating Mini-TOC..."}
            update_mini_toc(
                paths["toc_file"],
                filename,
                format_toc_title(dt),
                release_date=dt,
            )
            yield {"type": "step", "index": i, "step": "toc_done", "message": "Updated Mini-TOC"}

            # Page complete
            yield {
                "type": "page_done",
                "index": i,
                "filename": filename,
                "message": f"Complete: {filename}",
            }
            imported_count += 1

        except Exception as e:
            logger.exception("Error importing page %s", page_id)
            yield {"type": "error", "index": i, "message": str(e)}
            error_count += 1

    # Final summary
    yield {
        "type": "complete",
        "imported": imported_count,
        "skipped": skipped_count,
        "errors": error_count,
        "message": f"Import complete: {imported_count} imported, {skipped_count} skipped, {error_count} error(s)",
    }

File: ConfluenceToFlare/ConfluenceToFlare/Services/ImportEngine.swift
import Foundation
import os

/// Orchestrates the full import pipeline, yielding progress events via AsyncStream.
///
/// Mirrors the Python import_engine.py generator, replacing `yield` with
/// `AsyncStream.Continuation.yield`.
struct ImportEngine {
    private static let logger = Logger(subsystem: "ConfluenceToFlare", category: "ImportEngine")
    private static let manifestFilename = ".import_manifest.json"

    struct ManifestEntry: Codable {
        let confluencePageID: String
        let confluenceTitle: String
        let importedAt: String

        enum CodingKeys: String, CodingKey {
            case confluencePageID = "confluence_page_id"
            case confluenceTitle = "confluence_title"
            case importedAt = "imported_at"
        }
    }

    static func loadImportManifest(releaseNotesDir: URL) -> [String: ManifestEntry] {
        let manifestPath = releaseNotesDir.appendingPathComponent(manifestFilename)
        guard let data = try? Data(contentsOf: manifestPath),
              let manifest = try? JSONDecoder().decode([String: ManifestEntry].self, from: data) else {
            return [:]
        }
        return manifest
    }

    static func saveImportManifest(_ manifest: [String: ManifestEntry], releaseNotesDir: URL) {
        let manifestPath = releaseNotesDir.appendingPathComponent(manifestFilename)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(manifest) else { return }
        try? data.write(to: manifestPath)
    }

    static func getImportedPageIDs(releaseNotesDir: URL) -> Set<String> {
        let manifest = loadImportManifest(releaseNotesDir: releaseNotesDir)
        return Set(manifest.values.map(\.confluencePageID))
    }

    /// Get filenames of existing release note HTM files in the Flare project.
    static func getExistingReleases(releaseNotesDir: URL) -> Set<String> {
        let fm = FileManager.default
        guard fm.fileExists(atPath: releaseNotesDir.path) else { return Set() }

        do {
            let files = try fm.contentsOfDirectory(at: releaseNotesDir, includingPropertiesForKeys: nil)
            return Set(
                files.map(\.lastPathComponent)
                    .filter { $0.hasPrefix("Release Notes ") && $0.hasSuffix(".htm") }
            )
        } catch {
            return Set()
        }
    }

    /// Run the import pipeline, returning an AsyncStream of progress events.
    ///
    /// - Parameters:
    ///   - pageIDs: List of Confluence page IDs to import.
    ///   - client: Authenticated Confluence client.
    ///   - settings: App settings with Flare project paths.
    ///   - force: If true, overwrite existing Flare files.
    /// - Returns: AsyncStream of ImportEvent values.
    static func runImport(
        pageIDs: [String],
        client: ConfluenceClient,
        settings: AppSettings,
        force: Bool = false
    ) -> AsyncStream<ImportEvent> {
        AsyncStream { continuation in
            Task {
                let total = pageIDs.count
                continuation.yield(.start(total: total, message: "Starting import of \(total) page(s)"))

                // Validate paths
                guard let projectRoot = settings.resolvedProjectRoot,
                      FileManager.default.fileExists(atPath: projectRoot.path) else {
                    continuation.yield(.error(index: -1, message: "Flare project not found at \(settings.flareProjectRoot)"))
                    continuation.yield(.complete(imported: 0, skipped: 0, errors: 1, message: "Import failed"))
                    continuation.finish()
                    return
                }

                guard let releaseNotesDir = settings.resolvedReleaseNotesDir(),
                      let imagesDir = settings.resolvedImagesDir(),
                      let overviewFile = settings.resolvedOverviewFile(),
                      let tocFile = settings.resolvedTocFile() else {
                    continuation.yield(.error(index: -1, message: "Could not resolve Flare project paths"))
                    continuation.yield(.complete(imported: 0, skipped: 0, errors: 1, message: "Import failed"))
                    continuation.finish()
                    return
                }

                var manifest = loadImportManifest(releaseNotesDir: releaseNotesDir)
                let importedPageIDs = getImportedPageIDs(releaseNotesDir: releaseNotesDir)

                var importedCount = 0
                var skippedCount = 0
                var errorCount = 0

                for (i, pageID) in pageIDs.enumerated() {
                    // Check for cancellation
                    if Task.isCancelled {
                        continuation.yield(.error(index: -1, message: "Import cancelled"))
                        break
                    }

                    do {
                        // Fetch title
                        let title = try await client.getPageTitle(pageID: pageID)
                        continuation.yield(.pageStart(index: i, pageID: pageID, title: title, message: "Importing: \(title)"))

                        // Parse date
                        guard let dt = DateParser.parseConfluenceTitle(title) else {
                            continuation.yield(.error(index: i, message: "Could not parse date from title: \(title)"))
                            errorCount += 1
                            continue
                        }

                        let filename = DateParser.formatFlareFilename(dt)
                        let displayDate = DateParser.formatDisplayDate(dt)
                        let imageFolderName = DateParser.formatImageFolder(dt)

                        // Check if already imported (by Confluence page ID)
                        if importedPageIDs.contains(pageID) && !force {
                            continuation.yield(.skip(index: i, filename: filename, message: "Already exists: \(filename)"))
                            skippedCount += 1
                            continue
                        }

                        // Step 1: Download images
                        continuation.yield(.step(index: i, step: "images", message: "Downloading images..."))
                        let imageDest = try ImageHandler.createImageFolder(imagesBase: imagesDir, folderName: imageFolderName)
                        let imageMapping = try await ImageHandler.downloadAndPlaceImages(client: client, pageID: pageID, destFolder: imageDest)
                        let imgCount = imageMapping.count
                        continuation.yield(.step(index: i, step: "images_done", message: "Downloaded \(imgCount) image(s)"))

                        if Task.isCancelled { break }

                        // Step 2: Fetch and convert content
                        continuation.yield(.step(index: i, step: "content", message: "Converting page content..."))
                        let xhtml = try await client.getPageContent(pageID: pageID)
                        let bodyContent = try ContentConverter.convert(xhtml: xhtml, imageMapping: imageMapping, imageFolder: imageFolderName)
                        continuation.yield(.step(index: i, step: "content_done", message: "Converted page content"))

                        if Task.isCancelled { break }

                        // Step 3: Render template
                        continuation.yield(.step(index: i, step: "render", message: "Rendering Flare topic..."))
                        let htm = TemplateRenderer.render(displayDate: displayDate, bodyContent: bodyContent)
                        continuation.yield(.step(index: i, step: "render_done", message: "Rendered Flare topic"))

                        // Step 4: Write HTM file
                        continuation.yield(.step(index: i, step: "write", message: "Writing \(filename)..."))
                        let destFile = releaseNotesDir.appendingPathComponent(filename)
                        try FileManager.default.createDirectory(at: destFile.deletingLastPathComponent(), withIntermediateDirectories: true)
                        try BOMWriter.write(htm, to: destFile)
                        continuation.yield(.step(index: i, step: "write_done", message: "Wrote \(filename)"))

                        // Step 5: Update Overview page
                        continuation.yield(.step(index: i, step: "overview", message: "Updating Overview page..."))
                        try FlareUpdater.updateOverviewPage(
                            overviewPath: overviewFile,
                            filename: filename,
                            linkText: DateParser.formatOverviewLinkText(dt),
                            releaseDate: dt
                        )
                        continuation.yield(.step(index: i, step: "overview_done", message: "Updated Overview page"))

                        // Step 6: Update Mini-TOC
                        continuation.yield(.step(index: i, step: "toc", message: "Updating Mini-TOC..."))
                        try FlareUpdater.updateMiniToc(
                            tocPath: tocFile,
                            filename: filename,
                            tocTitle: DateParser.formatTocTitle(dt),
                            releaseDate: dt
                        )
                        continuation.yield(.step(index: i, step: "toc_done", message: "Updated Mini-TOC"))

                        // Page complete
                        continuation.yield(.pageDone(index: i, filename: filename, message: "Complete: \(filename)"))
                        importedCount += 1

                        // Record in import manifest
                        manifest[filename] = ManifestEntry(
                            confluencePageID: pageID,
                            confluenceTitle: title,
                            importedAt: ISO8601DateFormatter().string(from: Date())
                        )
                        ImportEngine.saveImportManifest(manifest, releaseNotesDir: releaseNotesDir)

                    } catch {
                        logger.error("Error importing page \(pageID): \(error.localizedDescription)")
                        continuation.yield(.error(index: i, message: error.localizedDescription))
                        errorCount += 1
                    }
                }

                // Final summary
                continuation.yield(.complete(
                    imported: importedCount,
                    skipped: skippedCount,
                    errors: errorCount,
                    message: "Import complete: \(importedCount) imported, \(skippedCount) skipped, \(errorCount) error(s)"
                ))
                continuation.finish()
            }
        }
    }
}
